### PR TITLE
Bug 1345409 - Ensure that we always close the selected tab last when …

### PR DESCRIPTION
**Pull request by fluffyemily**
…closing all private tabs

In top tabs, trying to close the selected tab causes a new tab to be chosen, but by the time the code comes to render it, it may no longer exist.
Ensuring we close the selected tab last means that the new selected tab will come from the non-private list and no crashing happens.

https://bugzilla.mozilla.org/show_bug.cgi?id=1345409

### CODE REVIEW

File: Client/Frontend/Browser/TabManager.swift [hunk 1]
```diff
@@ -414,7 +414,18 @@ class TabManager: NSObject {
     /// - Parameter notify: if set to true, the delegate is called when a tab is
     ///   removed.
     func removeAllPrivateTabsAndNotify(_ notify: Bool) {
-        privateTabs.forEach({ removeTab($0, flushToDisk: true, notify: notify) })
+        // if there is a selected tab, it needs to be closed last
+        // this is important for TopTabs as otherwise the selection of the new tab
+        // causes problems as it may no longer be present.
+        // without this we get a nasty crash
+        guard let selectedTab = selectedTab,
```
Comment (farhanpatel) [inline Client/Frontend/Browser/TabManager.swift]: Can't we avoid all the checks here? 
If selectTab doesnt exist then the filter at the bottom will just return all the private tabs. Thats fine, we still need to remove them. 
If selectTab isnt in privateTabs thats also okay. We still need to remove all the private tabs.

File: Client/Frontend/Browser/TabManager.swift [hunk 2]
```diff
@@ -414,7 +414,18 @@ class TabManager: NSObject {
...
+            let _ = privateTabs.index(of: selectedTab) else {
+            return privateTabs.forEach({ removeTab($0, flushToDisk: true, notify: notify) })
+        }
+
+        let nonSelectedTabsForRemoval = privateTabs.filter { $0 != selectedTab }
+        nonSelectedTabsForRemoval.forEach({ removeTab($0, flushToDisk: true, notify: notify) })
+        removeTab(selectedTab, flushToDisk: true, notify: notify)
```
Comment (farhanpatel) [inline Client/Frontend/Browser/TabManager.swift]: You could just a check here to check if selectTab is private and if it is. delete it and avoid having to check above.

**Comment by farhanpatel** [comment]
These are just some style changes. Not really that important. I'd rather have this in the beta going out today than to wait for this minor change. So I'm going to merge it in 😄

**Comment by fluffyemily** [comment]
OK. I will make the changes as a follow up bug.

Sorry, didn't realise you wanted this today!

**Pull request merged**